python/ray/data/impl/split.py

```diff
@@ -0,0 +1,1106 @@
+    # availability that's at the block's primary location.
+    allocate_blocks_to_primary_actors(
+        block_set, actor_bin_set, actors_by_node_id, overallocate=not equal)
+
+    # Round 2: Allocate blocks to the actor with the most allocation
+    # capacity that's at one of the block's secondary locations.
+    allocate_blocks_to_secondary_actors(
+        block_set, actor_bin_set, actors_by_node_id, overallocate=not equal)
+
+    if equal:
+        # TODO(Clark): Some Ray clusters won't have access to node
+        # ID resources (e.g. if their "node" primitive is a container); should
+        # we expose a custom resource hook here? This will probably require
+        # support from ray.experimental.get_object_locations().
+        # We could also fall back to relying on locality-aware
+        # scheduling in that case, but that could result in a lot of
+        # unnecessary data transfer if the cluster is busy.
+
+        # Round 3: Split the blocks, attempting to allocate the splits to
+        # actors that are located on nodes that already have the block.
+        split_blocks_over_actors(
+            block_set, actor_bin_set,
+            lambda node_id: actors_by_node_id[node_id],
+            lambda node_id: f"node:{addresses_by_node_id[node_id]}")
+
+        # Round 4: Split the blocks, allocating the splits to arbitrary
+        # nodes.
+        split_blocks_over_arbitrary_actors(block_set, actor_bin_set)
+
+        block_set.resolve_metadata_futures()
+
+        if blocks_per_split is None:
+            # TODO(Clark): Tweak this default.
+            blocks_per_split = max(
+                100, min(2, math.ceil(capacity / DEFAULT_ROWS_PER_BLOCK)))
+        # TODO(Clark): This will further truncate capacity % blocks_per_split
+        # rows from each block; since it's not important for the blocks to
+        # be exactly balanced within splits, we should allow for off-by-one
+        # blocks to minimize truncation. We truncate for now in order to
+        # make the rebalancing logic much simpler.
+        block_size = capacity // blocks_per_split
+
+        # Round 5: Split blocks larger than the target block size into
+        # target-sized blocks; merge blocks smaller than the target block
+        # size into target-sized blocks.
+        rebalance_blocks(block_set, actor_bin_set, block_size)
+
+        block_set.resolve_metadata_futures()
+    else:
+        # We don't want to split blocks, so we allocate the remaining blocks
+        # to arbitrary actors.
+        allocate_blocks_to_arbitrary_actors(block_set, actor_bin_set)
+
+    # Build output block lists.
+    block_lists = []
+    for actor in locality_hints:
+        allocated_blocks = actor_bin_set.get_bin(actor).allocated_blocks
+        block_lists.append(
+            BlockList([block.block for block in allocated_blocks],
+                      [block.meta for block in allocated_blocks]))
+    return block_lists
+
+
+class BlockData:
```

Comment:
hmm the helper class itself takes ~300 lines of code. I think with this many code it might not simplify but actually complicated the code. I believe removing this should result in less code.  (for reference the original split implementation is only 200 lines of code)

Reply:
part of your complexity is you want to pass state between functions. Another solution to it to create a stateful like `DataSetSplitter` which captures all the state, but no more additional abstractions over what we have.